haystack/finder.py

```diff
@@ -96,3 +98,164 @@ def get_answers_via_similar_questions(self, question: str, top_k_retriever: int
+            - "reader_f1": Average overlap between predicted answers and their corresponding correct answers
+            - "total_retrieve_time": Time retriever needed to retrieve documents for all questions
+            - "avg_retrieve_time": Average time needed to retrieve documents for one question
+            - "total_reader_time": Time reader needed to extract answer out of retrieved documents for all questions
+                                   where the correct document is among the retrieved ones
+            - "avg_reader_time": Average time needed to extract answer out of retrieved documents for one question
+            - "total_finder_time": Total time for whole pipeline
+
+        :param label_index: Elasticsearch index where labeled questions are stored
+        :type label_index: str
+        :param doc_index: Elasticsearch index where documents that are used for evaluation are stored
+        :type doc_index: str
+        :param top_k_retriever: How many documents per question to return and pass to reader
+        :type top_k_retriever: int
+        :param top_k_reader: How many answers to return per question
+        :type top_k_reader: int
+        """
+        finder_start_time = time.time()
+        # extract all questions for evaluation
+        filter = {"origin": label_origin}
+        questions = self.retriever.document_store.get_all_docs_in_index(index=label_index, filters=filter)
+
+        correct_retrievals = 0
+        correct_readings = 0
+        summed_avg_precision_retriever = 0
+        summed_avg_precision_reader = 0
+        exact_matches = 0
+        summed_f1 = 0
+        retrieve_times = []
+        read_times = []
+
+        # retrieve documents
+        questions_with_docs = []
+        retriever_start_time = time.time()
+        for q_idx, question in enumerate(questions):
+            question_string = question["_source"]["question"]
+            single_retrieve_start = time.time()
+            retrieved_docs = self.retriever.retrieve(question_string, top_k=top_k_retriever, index=doc_index)
+            retrieve_times.append(time.time() - single_retrieve_start)
+            for doc_idx, doc in enumerate(retrieved_docs):
+                # check if correct doc among retrieved docs
+                if doc.meta["doc_id"] == question["_source"]["doc_id"]:
+                    correct_retrievals += 1
+                    summed_avg_precision_retriever += 1 / (doc_idx + 1)
+                    questions_with_docs.append({
+                        "question": question,
+                        "docs": retrieved_docs,
+                        "correct_es_doc_id": doc.id})
+                    break
+        retriever_total_time = time.time() - retriever_start_time
+        number_of_questions = q_idx + 1
+
+        previous_top_k_per_candidate = self.reader.top_k_per_candidate
+        previous_return_no_answers = self.reader.return_no_answers
+        self.reader.top_k_per_candidate = 1
```

Comment:
The top k per candidate is tricky with a farm reader and not correctly handled within haystack currently.